cloudinit/config/cc_growpart.py

```diff
@@ -293,10 +300,111 @@ def devent2dev(devent):
     return dev
 
 
+def get_mapped_device(blockdev):
+    """Returns underlying block device for a mapped device.
+
+    If it is mapped, blockdev will usually take the form of
+    /dev/mapper/some_name
+
+    If blockdev is a symlink pointing to a /dev/dm-* device, return
+    the device pointed to. Otherwise, return None.
+    """
+    realpath = os.path.realpath(blockdev)
+    if realpath.startswith("/dev/dm-"):
+        LOG.debug("%s is a mapped device pointing to %s", blockdev, realpath)
+        return realpath
+    return None
+
+
+def is_encrypted(blockdev) -> bool:
+    """
+    Check if a device is an encrypted device. blockdev should have
+    a /dev/dm-* path.
+    """
+    if not subp.which("cryptsetup"):
+        LOG.debug("cryptsetup not found. Assuming no encrypted partitions")
+        return False
+    is_encrypted = False
+    with suppress(subp.ProcessExecutionError):
+        subp.subp(["cryptsetup", "status", blockdev])
```

Comment:
`cryptsetup` should only return 0 if it's actually encrypted. Any other rc will raise the ProcessExecutionError.

Are you saying we should log the return code, just in case there's a different problem with running cryptsetup?

Reply:
I guess what I'm thinking here two things:
 - Should we really only surpress a specific non-zero exit code and log any unexpected error messages and codes here
 - Also, running cryptsetup status on a thin volume group I created still returns 0 exit code, but it isn't an encrypted device, so a zero exit code may be a false positive in this case for any LVM setup?
```bash
csmith@downtown:~$ sudo cryptsetup status /dev/mapper/sru_lvmpool-LXDThinPool_tdata
/dev/mapper/sru_lvmpool-LXDThinPool_tdata is active and is in use.
  type:    n/a
csmith@downtown:~$ echo $?
0
```

Reply:
I'm wondering if we can use `dmsetup ls --target crypt` here to list known crypt devices.
On my vm with luks enabled for the root FS I get
```bash
$ dmsetup ls --target crypt
dm_crypt-0     (253, 0)
```

Reply:
The problem with `dmsetup ls` is that it's not easy to know how to tie values in that list to the specific device being resized.

In a previous iteration of this PR, I had a `cryptsetup isLuks` check. Based on a [misunderstanding of a comment](https://github.com/canonical/cloud-init/pull/1032#discussion_r738773375), I removed it. It looks like he was saying that we don't need to restrict resizing mapped devices to _only_ Luks containers, but I want to specifically do that right now as we haven't had support for LVM sizing, and I'd rather not add it here without additional testing. So I added back the `isLuks` check.

Reply:
+1 this context makes sense. And crypsetup isLuks behaves as expected on my non-luks volume group. I'm +1 on restricting this behavior to just encrypted partitions at the moment given the artifacts which allow for resizing. We can iterate toward LVM resizing in separate efforts.